### PR TITLE
GG-23175 Web Console should check IgniteFeatures instead of cluster version.

### PR DESCRIPTION
**Pull request by nva**

### CODE REVIEW

File: modules/web-console/frontend/app/types/Agent.ts [hunk 1]
```diff
@@ -69,3 +72,15 @@ export type VisorQueryResult =
 export type QuerySqlX2Response = VisorQueryResult
 
 export type QueryFetchFirstPageResult = VisorQueryResult
+
+export enum IgniteFeatures {
+	INDEXING= 15,
+	WC_ROLLING_UPGRADE_STATUS = 20,
+	WC_SNAPSHOT_CHAIN_MODE = 21
```
Comment (akuznetsov-gridgain) [inline modules/web-console/frontend/app/types/Agent.ts]: Where WC_DR_STATE_EVENTS ?

File: modules/web-console/web-console-common/src/main/java/org/apache/ignite/console/utils/Utils.java [hunk 2]
```diff
@@ -197,4 +198,26 @@ public static String extractErrorMessage(String prefix, Throwable e) {
...
     }
+
+    /**
+     * Checks that feature supported by cluster.
+     *
+     * @param featuresAttrBytes Byte array value of supported features.
+     * @param featureId Feature to check.
+     * @return {@code True} if feature is declared to be supported by cluster.
+     */
+    public static boolean gridSupports(byte[] featuresAttrBytes, int featureId) {
```
Comment (akuznetsov-gridgain) [inline modules/web-console/web-console-common/src/main/java/org/apache/ignite/console/utils/Utils.java]: May be `featureSupported` will be a better name than `gridSupports`?

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java [hunk 3]
```diff
@@ -82,7 +82,23 @@
     /**
      * A mode when data nodes throttle update rate regarding to DR sender load
      */
-    DR_DATA_NODE_SMART_THROTTLING(19);
+    DR_DATA_NODE_SMART_THROTTLING(19),
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java]: FEATURES NOT RESERVED AT:
https://ggsystems.atlassian.net/wiki/spaces/GG/pages/1192198276/Community+edition+features+list

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 4]
```diff
@@ -176,8 +177,16 @@ class ConnectionState {
 export default class AgentManager {
     static $inject = ['Demo', '$q', '$transitions', '$location', 'AgentModal', 'UserNotifications', 'IgniteVersion', 'ClusterLoginService'];
 
+    currentCluster$: Observable<ConnectionState>;
+    clusterIsActive$: Observable<boolean>;
+    clusterIsAvailable$: Observable<boolean>;
+
     clusterVersion: string;
 
+    indexing: boolean;
```
Comment (akuznetsov-gridgain) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: Do we need  `indexing`, `ruStaus` and `snapshoChainMode`? I think we should have only `supportedFeatures` and enum or bit numbers to later check with method like `isFeatureSupported(features, FEATURE)`

**Review by akuznetsov-gridgain — CHANGES_REQUESTED** [review]

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts [hunk 5]
```diff
@@ -2208,13 +2208,8 @@ export class NotebookCtrl {
...
             if (this._hasRunningQueries(paragraphs)) {
-                try {
-                    return Confirm.confirm('You have running queries. Are you sure you want to cancel them?')
-                        .then(() => this._closeOpenedQueries(paragraphs));
-                }
-                catch (err) {
-                    return Promise.reject(new CancellationError());
-                }
+                return Confirm.confirm('You have running queries. Are you sure you want to cancel them?')
+                    .then(() => this._closeOpenedQueries(paragraphs));
```
Comment (Klaster1) [inline modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: I don't see how this is related to the task. Commit messages tell nothing useful about this change either.

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 6]
```diff
@@ -118,8 +118,13 @@ const getLastActiveClusterId = () => {
 };
 
 class ConnectionState {
-    constructor(cluster) {
-        this.cluster = cluster;
+    cluster: AgentTypes.ClusterStats;
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]:
```suggestion
    cluster: AgentTypes.ClusterStats | null;
```

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 7]
```diff
@@ -165,9 +169,6 @@ class ConnectionState {
     }
 
     disconnect() {
-        if (this.cluster)
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: Why did you remove this?

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 8]
```diff
@@ -176,8 +177,16 @@ class ConnectionState {
 export default class AgentManager {
     static $inject = ['Demo', '$q', '$transitions', '$location', 'AgentModal', 'UserNotifications', 'IgniteVersion', 'ClusterLoginService'];
 
+    currentCluster$: Observable<ConnectionState>;
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: 👍

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts (hunk 4 above)
Comment (Klaster1) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: I agree with @akuznetsov-gridgain, there's too much going on AgentManager already, we can at least reduce the amount of properties it exposes.

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 9]
```diff
@@ -871,4 +890,42 @@ export default class AgentManager {
...
+    }
+
+    /**
+     * Decode base64 string to byte array.
+     *
+     * @param base64 Base64 string.
+     * @return {Uint8Array} Result byte array.
+     * @private
+     */
+    _base64ToArrayBuffer(base64: string) {
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]:
```suggestion
    _base64ToArrayBuffer(base64: string): Uint8Array {
```

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 10]
```diff
@@ -871,4 +890,42 @@ export default class AgentManager {
...
+        const bitIdx = feature & 0x7;
+
+        return (bytes[byteIdx] & (1 << bitIdx)) !== 0;
+    }
+
+    /**
+     * Decode base64 string to byte array.
+     *
+     * @param base64 Base64 string.
+     * @return {Uint8Array} Result byte array.
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]:
```suggestion
     * @return Result byte array.
```

**Review by Klaster1 — CHANGES_REQUESTED** [review]

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts (hunk 7 above)
Comment (nva) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: not found usage of this.cluster.disconnect

File: modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts (hunk 5 above)
Comment (nva) [reply modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts]: This code don't throw exceptions, just return promise.

File: modules/web-console/web-console-common/src/main/java/org/apache/ignite/console/utils/Utils.java (hunk 2 above)
Comment (nva) [reply modules/web-console/web-console-common/src/main/java/org/apache/ignite/console/utils/Utils.java]: Remove

**Review by vsisko — CHANGES_REQUESTED** [review]
A am agree with Aleksey and Ilya. Also please check that e2e tests are not broken.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java [hunk 11]
```diff
@@ -82,7 +82,23 @@
...
+
+    /**
+     * Support rolling upgrade status task for Web Console.
+     */
+    WC_ROLLING_UPGRADE_STATUS(20),
+
+    /**
+     * Support for chain parameter in snapshot delete task for Web Console.
+     */
+    WC_SNAPSHOT_CHAIN_MODE(21),
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java]: One more comment - you must add unit test for new features

**Review by akuznetsov-gridgain — CHANGES_REQUESTED** [review]

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts [hunk 12]
```diff
@@ -891,6 +885,21 @@ export default class AgentManager {
...
     }
 
+    /**
+     * Collect features supported by the cluster.
+     *
+     * @param cluster Cluster.
+     * @return all supported features.
+     */
+    allFeatures(cluster: AgentTypes.ClusterStats): AgentTypes.IgniteFeatures[] {
+        return _.reduce(AgentTypes.IgniteFeatures, (acc, featureId) => {
```
Comment (Klaster1) [inline modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: The new API looks awkward, but it's better than before. We should probably iterate on this later.

**Review by Klaster1 — APPROVED** [review]

**Review by Klaster1 — APPROVED** [review]

File: modules/web-console/frontend/app/types/Agent.ts (hunk 1 above)
Comment (nva) [reply modules/web-console/frontend/app/types/Agent.ts]: Will do this in separated ticket

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java (hunk 11 above)
Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteFeatures.java]: We already have e2e test

File: modules/web-console/frontend/app/types/Agent.ts (hunk 1 above)
Comment (Klaster1) [reply modules/web-console/frontend/app/types/Agent.ts]: @nva please make a task for this so we don't forget.

File: modules/web-console/frontend/app/modules/agent/AgentManager.service.ts (hunk 12 above)
Comment (nva) [reply modules/web-console/frontend/app/modules/agent/AgentManager.service.ts]: I tried to investigate this and found that I could not define an object with keys as in enumeration keys, only as values.

```
enum A {
  A1 = 10
}

type B = {
  [x in A]: boolean
}

const features: B = {
  10: true
}
```

**Review by akuznetsov-gridgain — APPROVED** [review]
